### PR TITLE
Add flow to fairseq

### PR DESCRIPTION
**Pull request by ilya-pchelintsev**

### CODE REVIEW

File: fairseq/fairseq/criterions/label_smoothed_cross_entropy.py [hunk 1]
```diff
@@ -65,15 +66,23 @@ def forward(self, model, sample, reduce=True):
         }
         return loss, sample_size, logging_output
```
Comment (pchelaa) [inline fairseq/fairseq/criterions/label_smoothed_cross_entropy.py]: А на `kl_loss` мы забили?

File: fairseq/fairseq/models/transformer.py [hunk 2]
```diff
@@ -285,6 +319,29 @@ def get_normalized_probs(
...
+
+        # return EncoderOut(
+        #     encoder_out=x,  # T x B x C
+        #     encoder_padding_mask=encoder_padding_mask,  # B x T
+        #     encoder_embedding=encoder_embedding,  # B x T x C
+        #     encoder_states=encoder_states,  # List[T x B x C]
+        # )
 
 @register_model("transformer_align")
 class TransformerAlignModel(TransformerModel):
```
Comment (pchelaa) [inline fairseq/fairseq/models/transformer.py]: `TransformerModel` мы переименовали в  `TransformerWithFlowModel `, так что тут NameError. Надо либо убрать `TransformerAlignModel`, либо вернуть `TransformerModel`. Я бы убрал `TransformerAlignModel`.

File: fairseq/fairseq/criterions/label_smoothed_cross_entropy.py (hunk 1 above)
Comment (valgger) [reply fairseq/fairseq/criterions/label_smoothed_cross_entropy.py]: я забыл просто добавить его к возвращаемым значениям, или можно к loss просто добавить

File: fairseq/fairseq/modules/transformer_layer.py [hunk 3]
```diff
@@ -257,15 +257,29 @@ def forward(
...
+        #     key=y,
+        #     value=y,
+        #     key_padding_mask=self_attn_padding_mask,
+        #     incremental_state=incremental_state,
+        #     need_weights=False,
+        #     attn_mask=self_attn_mask,
+        # )
+
+        # MY_CHANGES: output from decoder self-attention layer
+        x, attn, weights = self.self_attn(
```
Comment (pchelaa) [inline fairseq/fairseq/modules/transformer_layer.py]: `self.self_attn` никогда не возвращает три объекта. Видимо нужно просто убрать `attn`

File: fairseq/fairseq/modules/transformer_layer.py [hunk 4]
```diff
@@ -257,15 +257,29 @@ def forward(
...
             query=x,
             key=y,
             value=y,
             key_padding_mask=self_attn_padding_mask,
             incremental_state=incremental_state,
-            need_weights=False,
+            need_weights=True,
             attn_mask=self_attn_mask,
         )
+        z = torch.sum(weights * x, dim=0, keepdim=True)
```
Comment (pchelaa) [inline fairseq/fairseq/modules/transformer_layer.py]: Я не понял замысел этой строчки. И размерности `x` и `weights` тут совсем разные 😄

File: fairseq/fairseq/criterions/label_smoothed_cross_entropy_flow.py [hunk 5]
```diff
@@ -33,8 +33,9 @@ def forward(self, model, sample, reduce=True):
         2) the sample size, which is used as the denominator for the gradient
         3) logging outputs to display while training
         """
-        net_output = model(**sample['net_input'])
-        loss, nll_loss = self.compute_loss(model, net_output, sample, reduce=reduce)
+        encoder_output, decoder_output = model(**sample['net_input'])
+        loss, nll_loss = self.compute_rec_loss(model, decoder_output, sample, reduce=reduce)
+        kl_loss = self.compute_kl_loss(model, encoder_output, decoder_output)
```
Comment (pchelaa) [inline fairseq/fairseq/criterions/label_smoothed_cross_entropy_flow.py]: Тут надо семплировать `z` из `q(z|x,y)`

File: fairseq/fairseq/models/transformer_flow.py [hunk 6]
```diff
@@ -270,7 +304,7 @@ def forward(
             src_lengths=src_lengths,
             return_all_hiddens=return_all_hiddens,
         )
-        return decoder_out
+        return encoder_out, decoder_out
```
Comment (pchelaa) [inline fairseq/fairseq/models/transformer_flow.py]: А если так делать ничего не сломается на этапе инференса? Я пока не разобрался до конца, но подозреваю что сломается.

Comment (akirkov) [reply fairseq/fairseq/models/transformer_flow.py]: Конечно все сломается

File: fairseq/fairseq/models/transformer_flow.py [hunk 7]
```diff
@@ -27,14 +27,15 @@
...
 
 DEFAULT_MAX_SOURCE_POSITIONS = 1024
 DEFAULT_MAX_TARGET_POSITIONS = 1024
 
 
-@register_model("transformer")
-class TransformerModel(FairseqEncoderDecoderModel):
+@register_model("transformer_with_flow")
+class TransformerWithFlowModel(FairseqEncoderDecoderModel):
     """
```
Comment (akirkov) [inline fairseq/fairseq/models/transformer_flow.py:39]: А мы не хотим эту модель вообще в отдельный файл вынести, а исходный трансформер оставить неизменным?

File: fairseq/fairseq/modules/transformer_layer.py [hunk 8]
```diff
@@ -336,8 +316,8 @@ def forward(
                 ]
             else:
                 self_attn_state = [saved_state["prev_key"], saved_state["prev_value"]]
-            return x, attn, self_attn_state, z
-        return x, attn, None, z
```
Comment (akirkov) [inline fairseq/fairseq/modules/transformer_layer.py]: Мне кажется, этот файл на этом этапе вообще не нужно трогать.

**Comment by pchelaa** [comment]
Команды для запуска:
```
ln -s /Users/ilya/bitranslation/fairseq/fairseq_cli/train.py /usr/local/bin/fairseq-train
chmod +x /Users/ilya/bitranslation/fairseq/fairseq_cli/train.py
export PYTHONPATH=$PYTHONPATH:/Users/ilya/bitranslation/fairseq:/Users/ilya/bitranslation/flowseq
```

File: flowseq/flownmt/nnet/weightnorm.py [hunk 9]
```diff
@@ -19,9 +19,10 @@ def reset_parameters(self):
         self.linear = nn.utils.weight_norm(self.linear)
 
     def extra_repr(self):
-        return 'in_features={}, out_features={}, bias={}'.format(
-            self.in_features, self.out_features, self.bias is not None
-        )
+        # return 'in_features={}, out_features={}, bias={}'.format(
```
Comment (akirkov) [inline flowseq/flownmt/nnet/weightnorm.py:22]: Это зачем?

File: flowseq/flownmt/modules/priors/prior.py [hunk 10]
```diff
@@ -86,8 +86,8 @@ def decode(self, epsilon: torch.Tensor, tgt_mask: torch.Tensor,
         log_probs = log_probs.view(z.size(0), -1).sum(dim=1).mul(-0.5) + logdet
         return z, log_probs
 
-    def sample(self, nlengths: int, nsamples: int, src: torch.Tensor,
-               ctx: torch.Tensor, src_mask: torch.Tensor,
+    def sample(self, src: torch.Tensor,
+               src_mask: torch.Tensor,  nsamples = 1,
                tau=0.0, include_zero=False) -> Tuple[Tuple[torch.Tensor, torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor], Tuple[torch.Tensor, torch.Tensor, torch.Tensor]]:
```
Comment (akirkov) [inline flowseq/flownmt/modules/priors/prior.py]: А нам действительно нужно менять этот файл?

File: flowseq/flownmt/flows/couplings/coupling.py [hunk 11]
```diff
@@ -165,7 +165,9 @@ def init(self, data: torch.Tensor, mask: torch.Tensor, src: torch.Tensor, src_ma
 
     @overrides
     def extra_repr(self):
-        return 'inverse={}, in_channels={}, scale={}'.format(self.inverse, self.in_channels, self.scale)
+        # return 'inverse={}, in_channels={}, scale={}'.format(self.inverse, self.in_channels, self.scale)
```
Comment (akirkov) [inline flowseq/flownmt/flows/couplings/coupling.py:168]: Тот же вопрос. Мне казалось мы не будем менять код flowseq на данном этапе?

File: fairseq/fairseq/models/transformer_flow.py (hunk 7 above)
Comment (valgger) [reply fairseq/fairseq/models/transformer_flow.py:39]: Так и есть, мы вынесли в отдельный файл.
Исходный трансформер называется transformer.py - он неизменен.
Файл который мы редактируем называется transformer_flow.py.

File: flowseq/flownmt/flows/couplings/coupling.py (hunk 11 above)
Comment (valgger) [reply flowseq/flownmt/flows/couplings/coupling.py:168]: Не вижу смысле во флоусеке что-то сохранять в исходном виде. Мы же с ним ничего не будем сравнивать.
Вот с трансформером обычным будем поэтому он в неизменном виде.

File: flowseq/flownmt/modules/priors/prior.py (hunk 10 above)
Comment (valgger) [reply flowseq/flownmt/modules/priors/prior.py]: Его точно придется менять в методе sample. Так как в исходном флоусеке в это методе в начале предсказывается количество токенов, а потом генерится вектор z такой длины. Нам ничего предсказывать не надо, так что там точно нужно будет менять.

File: flowseq/flownmt/nnet/weightnorm.py (hunk 9 above)
Comment (valgger) [reply flowseq/flownmt/nnet/weightnorm.py:22]: это просто текстовое представление которое используется где-то в логах, если не закоментить, то выдает ошибку так-как не заданы некоторые параметры, например self.in_features.

Comment (valgger) [reply flowseq/flownmt/nnet/weightnorm.py:22]: Можно оставить, чтобы помнить что мы забыли задать какие-то параметры, но мне кажется они не не используются, а если будут использоваться то вылезет ошибка еще раз.

File: fairseq/fairseq/criterions/label_smoothed_cross_entropy_flow.py [hunk 12]
```diff
@@ -34,23 +34,21 @@ def forward(self, model, sample, reduce=True):
...
         net_output = model(**sample['net_input'])
-        loss, nll_loss = self.compute_loss(model, net_output, sample, reduce=reduce)
+        rec_loss, nll_loss = self.compute_rec_loss(model, net_output, sample, reduce)
+        kl_loss = self.compute_kl_loss(model, net_output, reduce)
+        loss = rec_loss + (kl_loss if kl_loss is not None else 0)
         sample_size = sample['target'].size(0) if self.args.sentence_avg else sample['ntokens']
         logging_output = {
             'loss': loss.data,
+            'rec_loss': rec_loss.data,
+            'kl_loss': kl_loss.data if kl_loss is not None else 0,
```
Comment (valgger) [inline fairseq/fairseq/criterions/label_smoothed_cross_entropy_flow.py:48]: Если хочешь чтобы оно выводилось при обучении, надо ниже в метод reduce_metrics тоже kl_loss вписать

File: fairseq/fairseq/models/transformer_flow.py [hunk 13]
```diff
@@ -794,21 +794,23 @@ def extract_features(
...
+                prev_output_tokens.eq(self.padding_idx),
+                encoder_out.encoder_out.transpose(0, 1),
+                encoder_out.encoder_padding_mask,
+                nsamples=1
+            )
+
+            z = z.squeeze(1)
+
+            prior_log_probs = self.prior.log_probability(
+                z,
```
Comment (valgger) [inline fairseq/fairseq/models/transformer_flow.py]: я бы попробовал, чтобы убедиться что все работает независимо. Для начала сюда класть константный вектор z фиксированной длины (лучше степени двойки, там с этим есть тонкий момент) и единичную маскую для него.

File: fairseq/fairseq/models/transformer_flow.py [hunk 14]
```diff
@@ -794,21 +794,23 @@ def extract_features(
...
-                )
+            z, posterior_log_probs = self.posterior.sample(
+                prev_output_tokens,
+                prev_output_tokens.eq(self.padding_idx),
+                encoder_out.encoder_out.transpose(0, 1),
+                encoder_out.encoder_padding_mask,
+                nsamples=1
+            )
+
+            z = z.squeeze(1)
```
Comment (valgger) [inline fairseq/fairseq/models/transformer_flow.py:849]: Это сломается если nsamples>1.
Во флоусеке делают так:
batch, _, length, nz = z.size()
z = z.view(-1, length, nz)

И еще надо не забыть ту же размерность логпробе прайора вернуть:
log_probs_prior = log_probs_prior.view(batch, n_samples)

File: fairseq/fairseq/criterions/label_smoothed_cross_entropy_flow.py [hunk 15]
```diff
@@ -65,7 +63,7 @@ def compute_kl_loss(self, model, net_output, reduce=True):
         posterior_log_probs = model.get_posterior_log_probability(net_output)
 
         if prior_log_probs is None or posterior_log_probs is None:
-            return 0
+            return None
 
         kl = (posterior_log_probs - prior_log_probs).mean(dim=1)
```
Comment (valgger) [inline fairseq/fairseq/criterions/label_smoothed_cross_entropy_flow.py]: это усреднение, mean(dim=1) - по батчу, должно быть еще одно по n_samples, иначе сломается при n_samples>1.

File: fairseq/fairseq/models/transformer_flow.py [hunk 16]
```diff
@@ -794,21 +794,23 @@ def extract_features(
...
+
+            prior_log_probs = self.prior.log_probability(
+                z,
+                torch.ones_like(prev_output_tokens),
+                encoder_out.encoder_out.transpose(0, 1),
+                encoder_out.encoder_padding_mask,
```
Comment (valgger) [inline fairseq/fairseq/models/transformer_flow.py]: Возможно эту маску надо инвертировать, ты не проверял?

File: fairseq/fairseq/models/transformer_flow.py [hunk 17]
```diff
@@ -794,21 +794,23 @@ def extract_features(
...
-                    prev_output_tokens.eq(self.padding_idx),
-                    encoder_out.encoder_out.transpose(0, 1),
-                    encoder_out.encoder_padding_mask,
-                    nsamples=1
-                )
+            z, posterior_log_probs = self.posterior.sample(
+                prev_output_tokens,
+                prev_output_tokens.eq(self.padding_idx),
+                encoder_out.encoder_out.transpose(0, 1),
+                encoder_out.encoder_padding_mask,
```
Comment (valgger) [inline fairseq/fairseq/models/transformer_flow.py]: Возможно эту маску нужно инвертировать, ты не проверял?
